sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/batch/PartitionScopeThresholds.java

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.cosmos.implementation.batch;
+
+import com.azure.cosmos.BulkProcessingOptions;
+import com.azure.cosmos.implementation.apachecommons.lang.tuple.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.util.function.Tuple2;
+
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.azure.cosmos.implementation.guava25.base.Preconditions.checkNotNull;
+
+public class PartitionScopeThresholds<TContext> {
+    private final static Logger logger = LoggerFactory.getLogger(PartitionScopeThresholds.class);
+
+    private final String pkRangeId;
+    private final BulkProcessingOptions<TContext> options;
+    private final AtomicInteger targetMicroBatchSize;
+    private final AtomicLong totalOperationCount;
+    private final AtomicReference<CurrentIntervalThresholds> currentThresholds;
+    private final String identifier = UUID.randomUUID().toString();
+    private final double minRetryRate;
+    private final double maxRetryRate;
+    private final double avgRetryRate;
+
+    public PartitionScopeThresholds(String pkRangeId, BulkProcessingOptions<TContext> options) {
+        checkNotNull(pkRangeId, "expected non-null pkRangeId");
+        checkNotNull(options, "expected non-null options");
+
+        this.pkRangeId = pkRangeId;
+        this.options = options;
+        this.targetMicroBatchSize = new AtomicInteger(options.getMaxMicroBatchSize());
+        this.totalOperationCount = new AtomicLong(0);
+        this.currentThresholds = new AtomicReference<>(new CurrentIntervalThresholds());
+        this.minRetryRate = Math.min(
+            options.getMinTargetedMicroBatchRetryRate(),
+            options.getMaxTargetedMicroBatchRetryRate());
```

Comment:
I think we checked minRetryRate should <= maxRetryRate in option, do we need here?

Reply:
Not anymore :-)

Reply:
Fixed

Reply:
Now understood what you meant - Fixed.